File: desktop/src/Qlipq.Ffmpeg/Estimate.cs
using System.Globalization;
using System.Text.RegularExpressions;
using Qlipq.Core;

namespace Qlipq.Ffmpeg;

public sealed record SizeEstimate
{
    public double Bytes { get; set; }

    /// <summary>True when the figure is a quality-model ballpark (CRF/preset), not a hard target.</summary>
    public bool Approximate { get; set; }
}

public static partial class Estimate
{
    /// <summary>
    /// Baseline bits-per-pixel (per frame) for each codec at CRF 23. Each 6 CRF steps roughly
    /// halves/doubles bitrate, so <c>bpp = base * 2^((23 - crf) / 6)</c>. Deliberately rough.
    /// </summary>
    private static double BppAtCrf23(string codec) => codec switch
    {
        "libx264" => 0.095,
        "libx265" => 0.06,
        _ => 0.095,
    };

    [GeneratedRegex(@"^[+-]?\d+")]
    private static partial Regex LeadingIntRegex();

    private static bool Truthy(int? value) => value is { } v && v != 0;

    /// <summary>Estimate the exported file size for a clip under the resolved encode settings.</summary>
    public static SizeEstimate EstimateExportSize(MediaInfo media, EditSpec spec, ResolvedEncode encode)
    {
        var duration = EditSpecs.EffectiveDuration(spec, media);
        if (duration <= 0) return new SizeEstimate { Bytes = 0, Approximate = false };

        var video = encode.Video;
        var forcedReencode = spec.Crop is not null || Truthy(video.ScaleHeight) || Truthy(video.Fps);
        var reencoding = encode.Reencode || forcedReencode;

        // Pure stream-copy: output ≈ the source scaled by the fraction of duration kept.
        if (!reencoding)
        {
            var sourceDuration = media.DurationSec != 0 ? media.DurationSec : duration;
            double sourceSize = media.SizeBytes ?? 0;
            return new SizeEstimate { Bytes = sourceSize * (duration / sourceDuration), Approximate = false };
        }

        // Output frame dimensions after crop + downscale.
        double cropW = spec.Crop?.Width ?? media.Width;
        double cropH = spec.Crop?.Height ?? media.Height;
        var outW = cropW;
        var outH = cropH;
        if (Truthy(video.ScaleHeight) && cropH > 0)
        {
            outH = video.ScaleHeight!.Value;
            outW = Math.Round(cropW * (video.ScaleHeight.Value / cropH) / 2.0, MidpointRounding.AwayFromZero) * 2;
        }
        var outFps = video.Fps is { } f && f > 0 ? f : media.Fps != 0 ? media.Fps : 30;

        var audioTracks = spec.AudioTracks.Count(t => t.Enabled);
        var audioKbps = audioTracks * ParseIntLeading(encode.Audio.Bitrate ?? "0");
        var audioBytes = audioKbps * 1000 * duration / 8;

        if (Truthy(video.BitrateKbps))
        {
            var bitrateBytes = (double)video.BitrateKbps!.Value * 1000 * duration / 8;
            return new SizeEstimate { Bytes = bitrateBytes + audioBytes, Approximate = false };
        }

        var baseBpp = BppAtCrf23(video.Codec ?? "libx264");
        var bpp = baseBpp * Math.Pow(2, (23 - (video.Crf ?? 20)) / 6.0);
        var videoBps = outW * outH * outFps * bpp;
        // Constrained VBR caps the bitrate, so the estimate can't exceed the cap.
        if (Truthy(video.MaxrateKbps)) videoBps = Math.Min(videoBps, (double)video.MaxrateKbps!.Value * 1000);
        var videoBytes = videoBps * duration / 8;
        return new SizeEstimate { Bytes = videoBytes + audioBytes, Approximate = true };
    }

    /// <summary>Mimics JS <c>parseInt(s, 10)</c>: leading integer, ignoring a trailing unit like <c>k</c>.</summary>
    private static double ParseIntLeading(string s)
    {
        var match = LeadingIntRegex().Match(s.Trim());
        return match.Success ? double.Parse(match.Value, CultureInfo.InvariantCulture) : double.NaN;
    }
}

File: iced/crates/qlipq-ffmpeg/src/estimate.rs
use std::sync::LazyLock;

use regex::Regex;

use qlipq_core::edit_spec::{effective_duration, EditSpec};
use qlipq_core::media::MediaInfo;

use crate::args::ResolvedEncode;

#[derive(Debug, Clone, PartialEq)]
pub struct SizeEstimate {
    pub bytes: f64,
    /// True when the figure is a quality-model ballpark (CRF/preset), not a hard target.
    pub approximate: bool,
}

static LEADING_INT: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^[+-]?[0-9]+").unwrap());

/// Baseline bits-per-pixel (per frame) at CRF 23; `bpp = base * 2^((23 - crf) / 6)`.
fn bpp_at_crf23(codec: &str) -> f64 {
    match codec {
        "libx264" => 0.095,
        "libx265" => 0.06,
        _ => 0.095,
    }
}

fn truthy(value: Option<i64>) -> bool {
    matches!(value, Some(v) if v != 0)
}

/// Estimate the exported file size for a clip under the resolved encode settings.
pub fn estimate_export_size(media: &MediaInfo, spec: &EditSpec, encode: &ResolvedEncode) -> SizeEstimate {
    let duration = effective_duration(spec, media);
    if duration <= 0.0 {
        return SizeEstimate { bytes: 0.0, approximate: false };
    }

    let video = &encode.video;
    let forced_reencode = spec.crop.is_some() || truthy(video.scale_height) || truthy(video.fps);
    let reencoding = encode.reencode || forced_reencode;

    // Pure stream-copy: output ≈ the source scaled by the fraction of duration kept.
    if !reencoding {
        let source_duration = if media.duration_sec != 0.0 { media.duration_sec } else { duration };
        let source_size = media.size_bytes.unwrap_or(0) as f64;
        return SizeEstimate { bytes: source_size * (duration / source_duration), approximate: false };
    }

    // Output frame dimensions after crop + downscale.
    let crop_w = spec.crop.as_ref().map(|c| c.width).unwrap_or(media.width) as f64;
    let crop_h = spec.crop.as_ref().map(|c| c.height).unwrap_or(media.height) as f64;
    let mut out_w = crop_w;
    let mut out_h = crop_h;
    if truthy(video.scale_height) && crop_h > 0.0 {
        let sh = video.scale_height.unwrap() as f64;
        out_h = sh;
        out_w = ((crop_w * (sh / crop_h)) / 2.0).round() * 2.0;
    }
    let out_fps = match video.fps {
        Some(f) if f > 0 => f as f64,
        _ => {
            if media.fps != 0.0 {
                media.fps
            } else {
                30.0
            }
        }
    };

    let audio_tracks = spec.audio_tracks.iter().filter(|t| t.enabled).count() as f64;
    let audio_kbps = audio_tracks * parse_int_leading(encode.audio.bitrate.as_deref().unwrap_or("0"));
    let audio_bytes = audio_kbps * 1000.0 * duration / 8.0;

    if truthy(video.bitrate_kbps) {
        let bitrate_bytes = video.bitrate_kbps.unwrap() as f64 * 1000.0 * duration / 8.0;
        return SizeEstimate { bytes: bitrate_bytes + audio_bytes, approximate: false };
    }

    let base = bpp_at_crf23(video.codec.as_deref().unwrap_or("libx264"));
    let bpp = base * 2_f64.powf((23 - video.crf.unwrap_or(20)) as f64 / 6.0);
    let mut video_bps = out_w * out_h * out_fps * bpp;
    if truthy(video.maxrate_kbps) {
        video_bps = video_bps.min(video.maxrate_kbps.unwrap() as f64 * 1000.0);
    }
    let video_bytes = video_bps * duration / 8.0;
    SizeEstimate { bytes: video_bytes + audio_bytes, approximate: true }
}

/// Mimics JS `parseInt(s, 10)`: leading integer, ignoring a trailing unit like `k`.
fn parse_int_leading(s: &str) -> f64 {
    LEADING_INT
        .find(s.trim())
        .and_then(|m| m.as_str().parse::<f64>().ok())
        .unwrap_or(f64::NAN)
}
